# 清理与分析数据

## 收集

#### 导入需要的库

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

from IPython.display import display
pd.set_option('display.max_rows',None)
pd.set_option('display.max_columns',None)
pd.set_option('max_colwidth',100)

#### twitter_archive_enhanced 保存为DataFrame。

In [2]:
df_raw = pd.read_csv('twitter-archive-enhanced.txt',sep=',',header=0)

#### image-prediction.tsv 转换为DataFrame

In [3]:
df_img = pd.read_csv('image-predictions.tsv',sep='\t',header=0)

#### tweet_json.txt 转换为DataFrame格式
+ 出现错误：'ValueError: Trailing data'
+ 解决方法：添加参数line=True
+ 参考：https://stackoverflow.com/questions/30088006/loading-a-file-with-more-than-one-line-of-json-into-pandas


In [4]:
df_supl = pd.read_json('tweet_json.txt',lines=True)

## 评估

### 针对twitter-archive-enhanced.csv文件的评估

In [5]:
df_raw.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1725  680070545539371008                    NaN                  NaN   
414   822859134160621569                    NaN                  NaN   
359   829878982036299777                    NaN                  NaN   
262   842765311967449089                    NaN                  NaN   
2145  669926384437997569                    NaN                  NaN   

                      timestamp  \
1725  2015-12-24 17:00:27 +0000   
414   2017-01-21 17:31:20 +0000   
359   2017-02-10 02:25:42 +0000   
262   2017-03-17 15:51:22 +0000   
2145  2015-11-26 17:11:11 +0000   

                                                                                   source  \
1725  <a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>   
414    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
359    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
262    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2145   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                     text  \
1725  Say hello to Emmie. She's trapped in an ornament. Tragic af. Looks pretty content tho. Maybe it'...   
414   This is Hobbes. He was told he was going to the park. Ended up at the vet. H*ckin bamboozled. Qu...   
359   RT @dog_rates: This is Loki. He smiles like Elvis. Ain't nothin but a hound doggo. 12/10 https:/...   
262   Meet Indie. She's not a fan of baths but she's definitely a fan of hide &amp; seek. 12/10 click ...   
2145   I've never seen a dog so genuinely happy about a tennis ball. 12/10 s'cute https://t.co/9RYY2NtHDw   

      retweeted_status_id  retweeted_status_user_id  \
1725                  NaN                       NaN   
414                   NaN                       NaN   
359          8.269587e+17              4.196984e+09   
262                   NaN                       NaN   
2145                  NaN                       NaN   

     retweeted_status_timestamp  \
1725                        NaN   
414                         NaN   
359   2017-02-02 01:01:21 +0000   
262                         NaN   
2145                        NaN   

                                                                                            expanded_urls  \
1725                                      https://twitter.com/dog_rates/status/680070545539371008/photo/1   
414                                       https://twitter.com/dog_rates/status/822859134160621569/photo/1   
359   https://twitter.com/dog_rates/status/826958653328592898/photo/1,https://twitter.com/dog_rates/st...   
262   https://www.gofundme.com/get-indie-home/,https://twitter.com/dog_rates/status/842765311967449089...   
2145                                      https://twitter.com/dog_rates/status/669926384437997569/photo/1   

      rating_numerator  rating_denominator    name  doggo floofer pupper puppo  
1725                 9                  10   Emmie   None    None   None  None  
414                 12                  10  Hobbes   None    None   None  None  
359                 12                  10    Loki  doggo    None   None  None  
262                 12                  10   Indie   None    None   None  None  
2145                12                  10    None   None    None   None  None

In [6]:
df_raw.shape

(2356, 17)

In [7]:
df_raw.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator',
       'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo'],
      dtype='object')

#### 共有17个特征，2356条记录。

+ tweet_id，是每个tweet的唯一标识
+ 'in_reply_to_status_id',Nullable. If the represented Tweet is a reply, this field will contain the integer representation of the original Tweet’s ID
+ 'in_reply_to_user_id', Nullable. If the represented Tweet is a reply, this field will contain the integer representation of the original Tweet’s author ID. This will not necessarily always be the user directly mentioned in the Tweet.
+ 'timestamp',
+ 'source',Utility used to post the Tweet, as an HTML-formatted string. 
+ 'text',The actual UTF-8 text of the status update.tweet中的文本内容
+ 'retweeted_status_id',
+ 'retweeted_status_user_id',
+ 'retweeted_status_timestamp', 
+ 'expanded_urls',
+ 'rating_numerator',评分的分子部分，一般大于10，也可能存在小于10的情况。
+ 'rating_denominator',评分的分母部分，一般为10
+ 'name',来自tweet的文本内容，判断是狗的名字。
+ 'doggo',一种狗的等级，等级一共有四种。
+ 'floofer',一种狗的等级，等级一共有四种。
+ 'pupper',一种狗的等级，等级一共有四种。
+ 'puppo'，一种狗的等级，等级一共有四种。

参考：
+ https://developer.twitter.com/en/docs/tweets/data-dictionary/overview/tweet-object

In [8]:
df_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [9]:
df_raw.isnull().any()

tweet_id                      False
in_reply_to_status_id          True
in_reply_to_user_id            True
timestamp                     False
source                        False
text                          False
retweeted_status_id            True
retweeted_status_user_id       True
retweeted_status_timestamp     True
expanded_urls                  True
rating_numerator              False
rating_denominator            False
name                          False
doggo                         False
floofer                       False
pupper                        False
puppo                         False
dtype: bool

In [10]:
df_raw.duplicated(subset=['tweet_id','name']).sum()

0

In [11]:
df_raw['name'].value_counts().nlargest(20)

None       745
a           55
Charlie     12
Cooper      11
Oliver      11
Lucy        11
Penny       10
Lola        10
Tucker      10
Bo           9
Winston      9
Sadie        8
the          8
an           7
Toby         7
Daisy        7
Buddy        7
Bailey       7
Scout        6
Jack         6
Name: name, dtype: int64

In [12]:
df_raw[['rating_numerator','rating_denominator']].describe()

rating_numerator  rating_denominator
count       2356.000000         2356.000000
mean          13.126486           10.455433
std           45.876648            6.745237
min            0.000000            0.000000
25%           10.000000           10.000000
50%           11.000000           10.000000
75%           12.000000           10.000000
max         1776.000000          170.000000

In [13]:
df_raw.rating_numerator.value_counts().nlargest(20)

12     558
11     464
10     461
13     351
9      158
8      102
7       55
14      54
5       37
6       32
3       19
4       17
1        9
2        9
420      2
0        2
15       2
75       2
80       1
20       1
Name: rating_numerator, dtype: int64

In [14]:
df_raw.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [15]:
for i in df_raw.text.values[:5]:
    display(i) 

"This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU"

"This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV"

'This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB'

'This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ'

'This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f'

In [16]:
df_raw.query("name == 'a'").text.values[:5]

array(['Here is a pupper approaching maximum borkdrive. Zooming at never before seen speeds. 14/10 paw-inspiring af \n(IG: puffie_the_chow) https://t.co/ghXBIIeQZF',
       'Here is a perfect example of someone who has their priorities in order. 13/10 for both owner and Forrest https://t.co/LRyMrU7Wfq',
       'Guys this is getting so out of hand. We only rate dogs. This is a Galapagos Speed Panda. Pls only send dogs... 10/10 https://t.co/8lpAGaZRFn',
       'This is a mighty rare blue-tailed hammer sherk. Human almost lost a limb trying to take these. Be careful guys. 8/10 https://t.co/TGenMeXreW',
       'Viewer discretion is advised. This is a terrible attack in progress. Not even in water (tragic af). 4/10 bad sherk https://t.co/L3U0j14N5R'],
      dtype=object)

In [17]:
df_raw[df_raw.rating_numerator==45].text.values

array(['From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 would hug all at once https://t.co/sedre1ivTK'],
      dtype=object)

In [18]:
df_raw[df_raw.rating_numerator==1776].text.values

array(["This is Atticus. He's quite simply America af. 1776/10 https://t.co/GRXwMxLBkh"],
      dtype=object)

In [19]:
df_raw[df_raw.rating_numerator==0].text.values[:]

array(["When you're so blinded by your systematic plagiarism that you forget what day it is. 0/10 https://t.co/YbEJPkg4Ag",
       "PUPDATE: can't see any. Even if I could, I couldn't reach them to pet. 0/10 much disappointment https://t.co/c7WXaB2nqX"],
      dtype=object)

In [20]:
df_raw[df_raw.rating_numerator==420].text.values[:1]

array(['@dhmontgomery We also gave snoop dogg a 420/10 but I think that predated your research'],
      dtype=object)

In [21]:
df_raw[df_raw.rating_numerator==20].text.values[:1]

array(["I'm aware that I could've said 20/16, but here at WeRateDogs we are very professional. An inconsistent rating scale is simply irresponsible"],
      dtype=object)

In [22]:
df_raw[df_raw.rating_denominator==170].text.values

array(['Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv'],
      dtype=object)

In [23]:
df_raw[(df_raw.in_reply_to_status_id.notnull())&(df_raw.in_reply_to_user_id.notnull())].shape

(78, 17)

In [24]:
df_raw[(df_raw.rating_numerator>20)&(df_raw.rating_denominator>10)].shape

(13, 17)

#### 评估df_raw小结：
+ 'timestamp'，'retweeted_status_timestamp'的数据类型是object
+ 'rating_numerator','rating_denominator',最大值分别为1776和170，分别大于均值13和10
+  'rating_numerator',存在较大的值，比如45，对应的'rating_denominator'是50，这是5只狗评分的分子和分母。
+ 'in_reply_to_status_id ','in_reply_to_user_id','retweeted_status_id ','retweeted_status_user_id','retweeted_status_timestamp','expanded_urls'存在null
+ 狗狗的等级分布在四个列中。
+ 有些狗没有等级分类。数值是None,不是null
+ 有些狗没有名字，数值是None,而不是null
+ 一些狗的名字是'an','the'或者'a'.
+ timestamp的类型是object

### 针对image-prediction.tsv文件的评估

In [25]:
df_img.sample(5)

tweet_id                                          jpg_url  \
1376  763183847194451968  https://pbs.twimg.com/media/CpdfpzKWYAAWSUi.jpg   
362   672834301050937345  https://pbs.twimg.com/media/CVZjOktVAAAtigw.jpg   
142   668614819948453888  https://pbs.twimg.com/media/CUdloW8WEAAxB_Y.jpg   
1473  780192070812196864  https://pbs.twimg.com/media/CtPMhwvXYAIt6NG.jpg   
1666  812709060537683968  https://pbs.twimg.com/media/C0dSk98WEAALyya.jpg   

      img_num                p1   p1_conf  p1_dog                p2   p2_conf  \
1376        1  miniature_poodle  0.354674    True        toy_poodle  0.338642   
362         1          Pembroke  0.582560    True          Cardigan  0.258869   
142         1           bustard  0.380772   False           pelican  0.100554   
1473        1            vizsla  0.144012    True          mongoose  0.091474   
1666        1      Irish_setter  0.326873    True  golden_retriever  0.182610   

      p2_dog        p3   p3_conf  p3_dog  
1376    True     teddy  0.155828   False  
362     True    nipple  0.033835   False  
142    False     crane  0.084713   False  
1473   False   hatchet  0.073545   False  
1666    True  Leonberg  0.156912    True

In [26]:
df_img.shape

(2075, 12)

In [27]:
df_img.columns

Index(['tweet_id', 'jpg_url', 'img_num', 'p1', 'p1_conf', 'p1_dog', 'p2',
       'p2_conf', 'p2_dog', 'p3', 'p3_conf', 'p3_dog'],
      dtype='object')

#### 共有12个特征，2075条记录。

+ tweet_id，是每个tweet的唯一标识
+ jpg_url 是预测的图像资源链接
+ img_num 最可信的预测结果对应的图像编号.
+ p1 是算法对推特中图片的一号预测 
+ p1_conf 是算法的一号预测的可信度
+ p1_dog 是一号预测该图片是否属于“狗”（有可能是其他物种，比如熊、马等）,True表示图片属于狗
+ p2 是算法对推特中图片预测的第二种可能性
+ p2_conf 是算法的二号预测的可信度
+ p2_dog 是二号预测该图片是否属于“狗”.
+ p3 图片预测的第三种可能性
+ p3_conf 三号预测的可信度
+ p3_dog 三号预测的图片是否属于狗。

参考：
+ udactiy

In [28]:
df_img.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [29]:
df_img.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [30]:
df_img.isna().any()

tweet_id    False
jpg_url     False
img_num     False
p1          False
p1_conf     False
p1_dog      False
p2          False
p2_conf     False
p2_dog      False
p3          False
p3_conf     False
p3_dog      False
dtype: bool

In [31]:
df_img.duplicated(subset=['tweet_id'],keep="first").sum()

0

In [32]:
df_img[(df_img.p1_conf>=df_img.p2_conf)&(df_img.p2_conf>=df_img.p3_conf)|(df_img.p1_conf>=df_img.p3_conf)].shape

(2075, 12)

In [33]:
df_img.img_num.value_counts()

1    1780
2     198
3      66
4      31
Name: img_num, dtype: int64

In [34]:
len(df_img.query("img_num >1"))

295

In [35]:
df_img.p1.str.istitle().sum()

402

In [36]:
df_img.p1.str.islower().sum()

1135

In [37]:
df_img[~((df_img.p1.str.istitle())|(df_img.p1.str.islower()))].shape

(538, 12)

#### 评估df_img小结：
+ p2,p2_conf,p3,p3_conf,p2_dog,p3_dog这些列，不需要。
+ img_num 表示对应的图片的编号，不需要这个列。
+ p1，狗的类型，存在大小写不一致的情况。



### 针对tweet_json.txt文件的评估

In [38]:
df_supl.sample(5)

contributors  coordinates          created_at display_text_range  \
298            NaN          NaN 2017-02-28 20:41:37           [0, 130]   
1623           NaN          NaN 2016-01-06 04:38:35           [0, 116]   
845            NaN          NaN 2016-08-17 20:27:34           [0, 113]   
2240           NaN          NaN 2015-11-21 02:07:05           [0, 139]   
1282           NaN          NaN 2016-03-11 21:15:02            [0, 78]   

                                                                                                 entities  \
298   {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 83667770191839...   
1623  {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 68459488491383...   
845   {'hashtags': [], 'symbols': [], 'user_mentions': [{'screen_name': 'ShopWeRateDogs', 'name': 'The...   
2240  {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 66788691771594...   
1282  {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [{'url': 'https://t.co/xdNNEZdGJY',...   

                                                                                        extended_entities  \
298   {'media': [{'id': 836677701918392320, 'id_str': '836677701918392320', 'indices': [131, 154], 'me...   
1623  {'media': [{'id': 684594884913836032, 'id_str': '684594884913836032', 'indices': [93, 116], 'med...   
845   {'media': [{'id': 766008557347233792, 'id_str': '766008557347233792', 'indices': [114, 137], 'me...   
2240  {'media': [{'id': 667886917715943426, 'id_str': '667886917715943426', 'indices': [116, 139], 'me...   
1282                                                                                                  NaN   

      favorite_count  favorited  \
298            13733      False   
1623            9807      False   
845             4120      False   
2240            2005      False   
1282           17430      False   

                                                                                                full_text  \
298   Say hello to Oliver. He's pretty exotic. Fairly pupset as well. Too many midterms coming pup. 11...   
1623  "FOR THE LAST TIME I DON'T WANNA PLAY TWISTER ALL THE SPOTS ARE GREY DAMN IT CINDY" ...10/10 htt...   
845   This is Angel. She stole the @ShopWeRateDogs shirt from her owner. Fits pretty well actually. 11...   
2240  This is Erik. He's fucken massive. But also kind. Let's people hug him for free. Looks soft. 11/...   
1282                       RT if you are as ready for summer as this pup is 12/10 https://t.co/xdNNEZdGJY   

      geo                  id              id_str in_reply_to_screen_name  \
298   NaN  836677758902222849  836677758902222848                    None   
1623  NaN  684594889858887680  684594889858887680                    None   
845   NaN  766008592277377025  766008592277377024                    None   
2240  NaN  667886921285246976  667886921285246976                    None   
1282  NaN  708400866336894977  708400866336894976                    None   

      in_reply_to_status_id  in_reply_to_status_id_str  in_reply_to_user_id  \
298                     NaN                        NaN                  NaN   
1623                    NaN                        NaN                  NaN   
845                     NaN                        NaN                  NaN   
2240                    NaN                        NaN                  NaN   
1282                    NaN                        NaN                  NaN   

      in_reply_to_user_id_str  is_quote_status lang place  possibly_sensitive  \
298                       NaN            False   en  None                 0.0   
1623                      NaN            False   en  None                 0.0   
845                       NaN            False   en  None                 0.0   
2240                      NaN            False   en  None                 0.0   
1282                      NaN            F

In [39]:
df_supl.shape

(2352, 31)

In [40]:
df_supl.columns

Index(['contributors', 'coordinates', 'created_at', 'display_text_range',
       'entities', 'extended_entities', 'favorite_count', 'favorited',
       'full_text', 'geo', 'id', 'id_str', 'in_reply_to_screen_name',
       'in_reply_to_status_id', 'in_reply_to_status_id_str',
       'in_reply_to_user_id', 'in_reply_to_user_id_str', 'is_quote_status',
       'lang', 'place', 'possibly_sensitive', 'possibly_sensitive_appealable',
       'quoted_status', 'quoted_status_id', 'quoted_status_id_str',
       'retweet_count', 'retweeted', 'retweeted_status', 'source', 'truncated',
       'user'],
      dtype='object')

#### 共有31个特征，2352条记录。

+ favorite_count，tweet被点赞的次数，反映阅读者对此tweet内容是否喜欢。
+ retweet_count，tweet被转发的次数，反映阅读者对此tweet内容喜欢的程度。
+ id,tweet的唯一标识


In [41]:
df_supl.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2352 entries, 0 to 2351
Data columns (total 31 columns):
contributors                     0 non-null float64
coordinates                      0 non-null float64
created_at                       2352 non-null datetime64[ns]
display_text_range               2352 non-null object
entities                         2352 non-null object
extended_entities                2073 non-null object
favorite_count                   2352 non-null int64
favorited                        2352 non-null bool
full_text                        2352 non-null object
geo                              0 non-null float64
id                               2352 non-null int64
id_str                           2352 non-null int64
in_reply_to_screen_name          78 non-null object
in_reply_to_status_id            78 non-null float64
in_reply_to_status_id_str        78 non-null float64
in_reply_to_user_id              78 non-null float64
in_reply_to_user_id_str          78 n

In [42]:
df_supl[['favorite_count','retweet_count']].describe()

favorite_count  retweet_count
count     2352.000000    2352.000000
mean      8109.198980    3134.932398
std      11980.795669    5237.846296
min          0.000000       0.000000
25%       1417.000000     618.000000
50%       3596.500000    1456.500000
75%      10118.000000    3628.750000
max     132318.000000   79116.000000

#### 评估df_img小结：
+ 保留favorite_count,retweet_count,id这两个列，其余列不需要。
+ 这个数据集是关于狗的类型，与df_raw都是存储狗这个observation。
+ id的数据类型是int64
+ 考虑是否从entities列中获取id，与现有的id列的数据进行验证



### 评估总结：

#### 质量

#####  **twitter-archive-enhanced数据**
+ 'in_reply_to_status_id','in_reply_to_user_id','retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp','expanded_urls'，'retweeted_status_timestamp',,'source'这些列不需要。
+ 有些狗没有名字，数值是None,'an','the'或者'a'.而不是null
+ 有些狗没有等级分类。数值是None,不是null
+ 'rating_numerator','rating_denominator',最大值分别为1776和170，分别大于均值13和10
+ timestamp的类型是object，tweet_id的类型是int64

##### **image-prediction数据**
+ p1，狗的类型，存在大小写不一致的情况。
+ p2,p2_conf,p3,p3_conf,p2_dog,p3_dog.img_num这些列，不需要。

##### **tweet_json数据**
+ 保留favorite_count,retweet_count,id这三个特征列，其余列不需要。


#### 整洁度
+ df_raw,df_img，df_supl的observation都是dog
+ df_raw中狗狗的等级分布在四个列中。


## 清理

#### 备份三个数据集

In [43]:
df_raw_clean = df_raw.copy()
df_img_clean = df_img.copy()
df_supl_clean = df_supl.copy()

#### 清理不需要的特征

##### 定义
+ 从df_raw删除下列不需要的列： 'in_reply_to_status_id','in_reply_to_user_id','retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp','expanded_urls'， 'timestamp'，'retweeted_status_timestamp','source'

+ 从df_img中删除p2,p2_conf,p3,p3_conf,p2_dog,p3_dog.img_num这些列

+ 从df_supl删除除了下列三个列'favorite_count','retweet_count','id'之外的其他列。

##### 代码

In [44]:
df_raw_clean.drop(columns=['in_reply_to_status_id','in_reply_to_user_id','retweeted_status_id','retweeted_status_user_id',
                            'retweeted_status_timestamp','expanded_urls','retweeted_status_timestamp','source'],inplace=True)
df_img_clean.drop(columns=['p2','p2_conf','p3','p3_conf','p2_dog','p3_dog','img_num'],inplace=True)
df_supl_clean = df_supl_clean[['id','favorite_count','retweet_count']]

##### 测试

In [45]:
df_raw_clean.columns

Index(['tweet_id', 'timestamp', 'text', 'rating_numerator',
       'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo'],
      dtype='object')

In [46]:
df_img_clean.columns

Index(['tweet_id', 'jpg_url', 'p1', 'p1_conf', 'p1_dog'], dtype='object')

In [47]:
df_supl_clean.columns

Index(['id', 'favorite_count', 'retweet_count'], dtype='object')

#### 清理空值

##### 定义
+ replace方法对df_raw中的'name'中的None,'a','an','the'替换为null
+ 使用dropna方法清理空值以及相应的记录。

##### 代码

In [48]:
df_raw_clean.name.replace(['None','a','an','the'],np.nan,inplace=True)
df_raw_clean.dropna(axis='index',inplace=True)

##### 测试

In [49]:
df_raw_clean.name.isnull().sum()

0

#### 清理数据类型

##### 定义
+ 使用astype或者to_datatime方法将df_raw中的'timestamp'的数据类型改为Datetime类型
+ 使用astype方法将df_raw,df_img中的'tweet_id'，以及df_supl中的'id'的数据类型改为str类型

##### 代码

In [51]:
df_raw_clean.timestamp = pd.to_datetime(df_raw_clean.timestamp)
df_raw_clean.tweet_id = df_raw_clean.tweet_id.astype(str)
df_img_clean.tweet_id = df_img_clean.tweet_id.astype(str)
df_supl_clean.id = df_supl_clean.id.astype(str)

##### 测试

In [52]:
df_raw_clean.dtypes

tweet_id                      object
timestamp             datetime64[ns]
text                          object
rating_numerator               int64
rating_denominator             int64
name                          object
doggo                         object
floofer                       object
pupper                        object
puppo                         object
dtype: object

In [53]:
df_img_clean.dtypes

tweet_id     object
jpg_url      object
p1           object
p1_conf     float64
p1_dog         bool
dtype: object

In [54]:
df_supl_clean.dtypes

id                object
favorite_count     int64
retweet_count      int64
dtype: object

### 存储清理后的主数据集

In [55]:
# 将清理后的主数据集保存为 csv 文件或者 SQLite 数据库文件


## 分析和可视化

> 提示：
- 在 Notebook 使用 pandas 或 SQL 分析主数据集，并生成至少三（3）个独立的结论。
- 在 Notebook 中使用 Python 绘图库或在 Tableau 中至少生成一（1）个可视化图表。


In [56]:
# 分析或可视化代码


与上面分析或可视化相关的见解或说明

In [57]:
# 分析或可视化代码


与上面分析或可视化相关的见解或说明

In [58]:
# 你需要添加更多的 code cell 和 markdown cell 来完成所有分析和可视化


更多说明或总结等

> 提示：在完成 Notebook 的所有内容之后，还需要完成两篇文本和图片组成的 PDF 报告。因为这两篇报告中只是文字和图片，不需要包含代码，你可以使用文字编辑软件，比如 Word 来完成：
- 创建一个 300-600 字的书面报告，命名为 `wrangle_report.pdf`，在该报告中简要描述你的数据整理过程。这份报告可以看作是一份内部文档，供你的团队成员查看交流。
- 创建一个 250 字以上的书面报告，命名为 `act_report.pdf`，在该报告中，你可以与读者交流观点，展示你使用整理过的数据生成的可视化图表。这份报告可以看作是一份外部文档，如博客帖子或杂志文章。